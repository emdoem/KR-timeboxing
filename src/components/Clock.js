import React from 'react';
import PropTypes from 'prop-types';

function Clock({ className, minutes, seconds }) {
    // here making sure the numbers display as double digits
    let minutesDisplay = minutes.toString();
    if (minutesDisplay.length < 2) {
        minutesDisplay = '0' + minutesDisplay;
    };
    let secondsDisplay = seconds.toString();
    if (secondsDisplay.length < 2) {
        secondsDisplay = '0' + secondsDisplay;
    };
    
    return <h2 className={"Clock " + className}>Pozosta≈Ço {minutesDisplay}:{secondsDisplay}</h2>
}
Clock.defaultProps = {
    className: ""
}
function NonNegativeNumbersType(props, propName, componentName) {
    if (props[propName] < 0) {
        return new Error(`Invalid prop '${propName}' issued to component '${componentName}'. It can't be negative.`);
    };
};
const NumberOrStringType = PropTypes.oneOfType([PropTypes.number, PropTypes.string]);
Clock.propTypes = {
    className: PropTypes.string.isRequired, 
    minutes: NumberOrStringType.isRequired, 
    seconds: NonNegativeNumbersType
}

export default Clock;