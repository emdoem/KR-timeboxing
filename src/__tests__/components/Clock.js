import React from "react";
import Clock from "../../components/Clock";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";

var root = null;
var clockRenderer = null;
describe('<Clock />', () => {
    xdescribe('when given minutes and seconds (DOM)', () => {
        beforeEach(() => {
            root = document.createElement("div");
            ReactDOM.render(
                <Clock minutes={10} seconds={20} />, root
            )
        })
        it('renders an h2 element', () => {            
            expect(root.childNodes[0].nodeName).toEqual("H2")            
        });
        it('sets a Clock className', () => {            
            expect(root.childNodes[0].className).toMatch(/Clock/)
        });  
        it('renders time properly', () => {            
            expect(root.childNodes[0].textContent).toMatch(/10:20/)
        });      
    });
    describe('when given minutes and seconds (TestRenderer)', () => {
        beforeEach(() => {
            clockRenderer = renderer.create(
                <Clock minutes={10} seconds={20} />    
            );                        
        })
        it('renders properly', () => {
            expect(clockRenderer.toJSON()).toMatchSnapshot();
            // console.log(clockRenderer.toJSON())
            // expect(clockRenderer.toJSON().type).toEqual("h2")
            // expect(clockRenderer.toJSON().props).toMatchObject({className: expect.stringMatching(/Clock/)})
            // expect(clockRenderer.toJSON().children).toEqual(expect.arrayContaining([  '10', '20' ]))

        })
        it('renders an h2 element', () => {            
            expect(clockRenderer.toJSON().type).toEqual("h2")
           
        });
        it('sets a Clock className', () => {            
            expect(clockRenderer.toJSON().props).toMatchObject({className: expect.stringMatching(/Clock/)})

        });  
        it('renders time properly', () => {            
            expect(clockRenderer.toJSON().children).toEqual(expect.arrayContaining([  '10', '20' ]))

        });      
    });
    

    it('sets className to empty string if not given anything else', () => {
        expect(<Clock minutes={10} seconds={20} />)
            .toEqual(<Clock className="" minutes={10} seconds={20} />)
    });
});