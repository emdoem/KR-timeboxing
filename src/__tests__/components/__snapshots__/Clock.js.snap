// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Clock /> when given minutes and seconds (TestRenderer) renders properly 1`] = `
<h2
  className="Clock "
>
  Pozosta≈Ço 
  10
  :
  20
</h2>
`;
